flowauth/backend/flowauth/roles.py

```diff
@@ -0,0 +1,162 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+from flask import Blueprint, jsonify, request, current_app
+import datetime
+
+from flask_login import current_user, login_required
+from flask_principal import Permission, RoleNeed
+
+from .models import Role, Scope, Server, User, db
+
+blueprint = Blueprint(__name__.split(".").pop(), __name__)
+
+admin_permission = Permission(RoleNeed("admin"))
+
+
+def role_to_dict(role):
+    return {
+        "id": role.id,
+        "name": role.name,
+        "scopes": sorted([scope.id for scope in role.scopes]),
+        "latest_token_expiry": role.latest_token_expiry.strftime(
+            "%Y-%m-%dT%H:%M:%S.%fZ"
+        ),
+        "longest_token_life_minutes": role.longest_token_life_minutes,
+        "server": role.server_id,
+        "users": sorted([user.id for user in role.users]),
+    }
+
+
+def roles_to_json(roles):
+    return jsonify(
+        sorted([role_to_dict(role) for role in roles], key=lambda x: x["id"])
+    )
+
+
+@blueprint.route("/", methods=["GET"])
+@login_required
+@admin_permission.require(http_exception=401)
+def list_roles():
+    return jsonify([role_to_dict(role) for role in Role.query.all()])
+
+
+@blueprint.route("/user/<user_id>")
+@login_required
+@admin_permission.require(http_exception=401)
+def list_user_roles(user_id):
+    user = User.query.filter(User.id == user_id).first_or_404()
+    return jsonify([role_to_dict(role) for role in user.roles])
+
+
+@blueprint.route("/<role_id>", methods=["GET"])
+@login_required
+@admin_permission.require(http_exception=401)
+def get_role(role_id):
+    role = Role.query.filter(Role.id == role_id).first()  # First or error?
+    return jsonify(role_to_dict(role))
+
+
+@blueprint.route("/", methods=["POST"])
```

Comment:
Now that roles are per-server (as opposed to groups, which could correspond to multiple servers), I wonder whether it would be more consistent to use `/roles/<server>` as the API route here (consistent with `/tokens/<server>`), instead of having the server ID in the json. This is not critical, though.

Reply:
That might cause confusion with the GET and PATCH methods, which are on `/roles/<role_id>`.

Reply:
Could change those to `/roles/<server_id>/<role_id>`, but I suppose that then requires remembering that it's `/roles/<server_id>/<role_id>` and not `/roles/<role_id>/<server_id>`. This change would make more sense if we were allowing role ID to be unique only per server (e.g. two different servers could each have an 'analyst' role), but if role ID alone is globally unique then having to specify both server_id and role_id in the route adds redundancy, so I think let's keep it as it is.